### PR TITLE
Vet clinic database: query and update animals table

### PR DESCRIPTION
**Pull request by Adams-Ghub**
# The following are the queries carried out and their results for the following tasks

## Task one
  - Inside a transaction update the animals table by setting the species column to unspecified. Verify that change was made.
  - Then roll back the change and verify that the species columns went back to the state before the transaction.
first_transaction_before_rollback 
first_transaction_after_rollback

## Task two
- Inside a transaction:
    - Update the animals table by setting the species column to digimon for all animals that have a name ending in mon.
    - Update the animals table by setting the species column to pokemon for all animals that don't have species already set.
    - Verify that changes were made.
    - Commit the transaction.
    - Verify that changes persist after commit.
secondTransactionCorrectedAfterCommit

## Task three
- Now, take a deep breath and... Inside a transaction delete all records in the animals table, then roll back the transaction.
- After the rollback verify if all records in the animals table still exists. After that, you can start breathing as usual 
thirdTransactionQueryCorrected
thirdTransactionCorrected

## Task four
- Inside a transaction:
    - Delete all animals born after Jan 1st, 2022.
    - Create a savepoint for the transaction.
    - Update all animals' weight to be their weight multiplied by -1.
    - Rollback to the savepoint
    - Update all animals' weights that are negative to be their weight multiplied by -1.
    - Commit transaction
ForthTransactioncCorrected

# Queries and their results to answer the following questions:

  ## How many animals are there?
 
how_many_animals_are_thereQuerry
how_many_animals_are_there

 ## How many animals have never tried to escape?
 how_many_animals_have_never_escapte

## What is the average weight of animals?
average_weight_of_animalsCorrected

 ## Who escapes the most, neutered or not neutered animals?
who_escapes_most_neutered_not_neutered

  ## What is the minimum and maximum weight of each type of animal?
min_and_max_weight_for_each_type

  ## What is the average number of escape attempts per animal type of those born between 1990 and 2000?
average_number_of_escape_from_1990_to_2000

### CODE REVIEW

File: data.sql [hunk 1]
```diff
@@ -4,3 +4,11 @@ INSERT INTO animals(name, date_of_birth, escape_attempt, neutered, weight_kg) VA
 INSERT INTO animals(name, date_of_birth, escape_attempt, neutered, weight_kg) VALUES('Gabumon', '2018-11-15',2, true,8) 
 INSERT INTO animals(name, date_of_birth, escape_attempt, neutered, weight_kg) VALUES('Pikachu', '2021-01-07',1, false, 15.04)
 INSERT INTO animals(name, date_of_birth, escape_attempt, neutered, weight_kg) VALUES('Devimon', '2017-05-12',5, true, 11)
+INSERT INTO animals (name, date_of_birth, escape_attempt, neutered, weight_kg) VALUES('Charmander', '2020-02-28',0, false,-11)
```
Comment (Munsa1) [inline data.sql]: - [ ] The date of birth for Charmander is `2020-02-08` and not `2020-02-28` as you have entered. Kindly fix this for accuracy.

File: queries.sql [hunk 2]
```diff
@@ -8,3 +8,54 @@ SELECT name,escape_attempt FROM animals WHERE weight_kg>10.5;
 SELECT * FROM animals WHERE neutered=true;
 SELECT * FROM animals WHERE name<>'Gabumon';
 SELECT * FROM animals WHERE weight_kg>=10.4 AND weight_kg<=17.3;
+
+-- First transaction block of codes
+BEGIN;
+UPDATE animals SET species = 'unspecified';
+ROLLBACK;
+COMMIT;
+
```
Comment (Munsa1) [inline queries.sql]: - [ ] In this first transaction we have to verify that the change is made to the species column of the animals table before and after the `ROLLBACK` using `SELECT species from animals;` as per requirements. Therefore your code sholud look something like:
e.g
```
   BEGIN; -- start transaction

   UPDATE animals
   SET species = 'unspecified'; -- make change

   SELECT species from animals; -- verify that change was made
   ROLLBACK;

   SELECT species from animals; -- verify that change was undone
```
###  Note that `COMMIT` statement is not required here

File: queries.sql [hunk 3]
```diff
@@ -8,3 +8,54 @@ SELECT name,escape_attempt FROM animals WHERE weight_kg>10.5;
...
+-- Second Transaction block of codes
+BEGIN;
+UPDATE animals SET species='digimon' WHERE name LIKE '%mon';
+UPDATE animals SET species ='pokemon' WHERE species IS NULL;
+COMMIT;
+
```
Comment (Munsa1) [inline queries.sql]: - [ ] Again here BEFORE and AFTER we commit the changes to the species table we must verify that te changes were made using the `SELECT` statement to see the change.

e.g 
``SELECT species from animals;``

File: queries.sql [hunk 4]
```diff
@@ -8,3 +8,54 @@ SELECT name,escape_attempt FROM animals WHERE weight_kg>10.5;
...
+-- Third Transaction block of codes
+BEGIN;
+DELETE FROM animals;
+ROLLBACK;
+COMMIT;
```
Comment (Munsa1) [inline queries.sql]: - [ ] Also here we have to verify the change was made BEFORE and AFTER the `ROLLBACK` statement by checking the count of the records in our animals table.

e.g
`SELECT COUNT(*) FROM ANIMALS;`


#### KIndly note that your screenshots have to be updated to reflect any relevant change made for these transactions from the first to the third transaction

File: queries.sql [hunk 5]
```diff
@@ -8,3 +8,54 @@ SELECT name,escape_attempt FROM animals WHERE weight_kg>10.5;
...
+
+-- how many animals are there?
+SELECT COUNT(*) AS total_animals FROM animals;
+
+-- how many animals have never tried to escape
+SELECT COUNT(*) AS no_escape FROM animals WHERE escape_attempt=0;
+
+-- what is the average weight on animals
+SELECT AVG(weight_kg) AS average_weight FROM animals;
+
```
Comment (Munsa1) [inline queries.sql:54]: - [ ] The expected result for this query should be `15.55`, your output is `15.0111`. While there isnt much of a big difference kindly fix this error as with large data sets these differences can becom significant and cause you to have incorrect results.

Your output
image

#### Note please update the screenshot for this query as well upon making changes

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @Adams-Ghub, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Provided Screenshots For Queries :heavy_check_mark: 
Correct Git Flow :heavy_check_mark: 
Descriptive Pr :heavy_check_mark:
 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql [hunk 6]
```diff
@@ -8,3 +8,61 @@ SELECT name,escape_attempt FROM animals WHERE weight_kg>10.5;
...
+COMMIT;
+
+-- how many animals are there?
+SELECT COUNT(*) AS total_animals FROM animals;
+
+-- how many animals have never tried to escape
+SELECT COUNT(*) AS no_escape FROM animals WHERE escape_attempt=0;
+
+-- what is the average weight on animals
+SELECT AVG(weight_kg) AS average_weight FROM animals;
```
Comment (Shedrack-Sunday) [inline queries.sql:53]: - [ ] You haven't updated the image screenshot of the changes asked by the previous reviewer to have the value as `15.55`. kindly update the screenshot

**Review by Shedrack-Sunday — CHANGES_REQUESTED** [review]
Hello @Adams-Ghub      👋 

# CHANGES REQUIRED ♻️

Great work on the project 👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯



## Highlights   
- Good use of GitHub features ✔️  
- Nice use SQL queries ✔️ 
  
# Required Changes  🌵 

- [ ] You made some efforts in the implementations which I must say is a job well done. However, there are some issues  I suggest   you make In line with the requirements 

_Check the comments for a detailed review 

# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by binhussen — APPROVED** [review]
Hi @Adams-Ghub , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

## Highlights
- Well Structured query ✔️
- Descriptive pull request screenshoots✔️
- Used Git flow ✔️
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**